### ISSUE TITLE
Post evaluate at different sizes

### ISSUE THREAD

**Issue by schrum2** [schrum2/MarioDiffusion#163]
This is related to #160 .

In general, we seem to be able to successfully train on datasets of varied sizes, which means we also need to evaluate the models at generating things at various sizes. This is already done to a very tiny extent. run_diffusion_multi.bat already generates both short (16) levels and long (128) but ignores scales in between. That is probably ok since those are unconditional models, but I figured I would point the issue out.

What I specifically want to happen before this issue is closed is for caption adherence of conditional models to take level size into account. When testing caption adherence using json data sets, the levels being generated by the prompts should have the same widths as the original levels, since that is a fair comparison. It's possible that a common solution could be used to address Issue #160, but that one has some extra complications.

**Comment by schrum2** [comment]
This isn't strictly the same issue, but I'm going to re-open this to request some related changes.

I want a batch file and instructions that simplifies training with the large mixed size samples.

1. Change the instructions in the README under "Datasets with longer levels" to mention that running `Mar1and2Mixed-data.bat` creates the mixed dataset by combining the results of widths 16, 32, 64, and 128.
2. I do not think that `train-conditional.bat` is capable of handling this large mixed dataset via a simple modification of command line parameters, but please correct me if I'm wrong. Assuming it is not possible, create `train-conditional-mixed.bat` to handle this. It will obviously be very similar to `train-conditional.bat`, but different in the conventions for the datasets it loads and the output names of directories.
3. For whichever approach works, add instructions in the README near the end (after your block embedding stuff, before "Tile based games besides Mario") that mentions how to train on the large dataset (include the "experimental" disclaimer).

Regarding Step 1 about creating the mixed data, please remind me: did GitHub complain about committing data that was that large? If not, we might commit it, but otherwise we'll leave it out.

Also, for the regular data that was already in the repo, I've updated it a few times based on small changes we've made (adding the word "tons" for example). Do your recent changes for RandomTest captions result in different data being created? Basically, whatever datasets are created by running the `Mar1and2-data.bat` command should be committed to the repo, so please do that as well if needed.

**Comment by fwilson12** [comment]
Github won't allow files over 100MB to be committed, and even after split_data is called on the large mixed dataset, the training split is nearly 200MB. This means users will need to locally create the datasets, which I've noted in the README. After the meeting with Dr. Kohls I'll recreate the regular data already in the repo and see if there are any changes

**Comment by fwilson12** [comment]
Just confirmed that the files generated by the recently updated RandomTest only differ from the old files due to the introduction of the "a ton" quantifier; all of the updated basic Mario datasets/tokenizers (everything generated by running Mar1and2.bat) are included in the recent commit.

**Comment by schrum2** [comment]
There is indeed a problem with the captions in RandomTest (or possibly with the tokenizer). I think the tokenizer is based on whatever is in the training data, so this must mean that the random caption generator generates tokens that never appear in the actual training data. What to do about that?

```
(SURF2026) C:\Users\he_de\GitHub\MarioDiffusion>python verify_data_and_tokenizer.py --json datasets\Mar1and2_16-32-64-128_RandomTest-regular.json --pkl datasets\Mar1and2Mixed_Tokenizer-regular.pkl
Dataset:   datasets\Mar1and2_16-32-64-128_RandomTest-regular.json  (100 entries)
Tokenizer: datasets\Mar1and2Mixed_Tokenizer-regular.pkl  (vocab size 48)
Captions found: 100

Total tokens scanned: 2384
Distinct tokens:      48

FAILURE: Found 3 illegal token(s) not present in the tokenizer vocabulary:

  'broken'  (appears 27 times)
      entry 2: one platform. a ton of coins. two rectangular block clusters. two irregular block clusters. many cannons. one broken pipe. giant gap with a few chunks of floor. several coin lines. two enemies. two walls.
      entry 3: a few coins. several enemies. one broken pipe. several coin lines. a few towers. one cannon. one rectangular block cluster. one descending staircase. two walls. several pipes. one upside down pipe.
      entry 5: two walls. ceiling with a few gaps. two coins. one pipe. a few loose blocks. two broken pipes. many enemies. two cannons. a few coin lines.
      entry 6: ceiling with two gaps. one broken pipe. a few coins. giant gap with several chunks of floor. one upside down pipe. several loose blocks. two coin lines. several rectangular block clusters.
      entry 7: one upside down pipe. one coin. a few ascending staircases. a few platforms. one broken pipe. two loose blocks. several coin lines. one enemy.
      entry 9: one broken pipe. a few question blocks. a ton of pipes. two cannons. several ascending staircases. one rectangular block cluster. a few enemies.
      entry 12: one descending staircase. ceiling with several gaps. giant gap with a few chunks of floor. a few coins. one coin line. a few ascending staircases. many cannons. two broken pipes. many loose blocks. one wall.
      entry 15: a ton of coins. two broken pipes. many towers. a few coin lines. several rectangular block clusters. two walls.
      entry 20: two pipes. one wall. one upside down pipe. giant gap with many chunks of floor. one broken pipe. two ascending staircases.
      entry 25: two broken pipes. two upside down pipes. two ascending staircases. a ton of coins. one pipe. two towers.
      entry 28: a few question blocks. several rectangular block clusters. two cannons. two broken pipes. two descending staircases.
      entry 33: a few coin lines. a few ascending staircases. one broken pipe. two cannons. many coins. a few rectangular block clusters.
      entry 35: two platforms. two irregular block clusters. floor with a ton of gaps. a few coin lines. one broken pipe. a few enemies. one cannon. one wall. two coins.
      entry 52: two walls. several ascending staircases. two platforms. one irregular block cluster. many cannons. two broken pipes. a few enemies. a few coin lines. a ton of coins. two loose blocks.
      entry 60: one enemy. a few coin lines. several coins. one descending staircase. two broken pipes. several irregular block clusters. several ascending staircases. many cannons. one tower.
      entry 62: a few coins. many upside down pipes. a few ascending staircases. several pipes. one question block. a few loose blocks. two broken pipes. many enemies. full floor.
      entry 65: many platforms. a ton of coins. a ton of coin lines. two walls. two broken pipes. full floor. two question blocks. one tower.
      entry 71: two walls. many question blocks. two coins. a few enemies. one broken pipe. one coin line. one loose block.
      entry 72: several towers. two rectangular block clusters. two broken pipes. a ton of coins. two upside down pipes. floor with several gaps. many coin lines. several enemies. full ceiling.
      entry 74: several upside down pipes. several coins. a few ascending staircases. two question blocks. one irregular block cluster. floor with many gaps. one broken pipe. several coin lines. a ton of enemies.
      entry 76: a ton of platforms. two cannons. many upside down pipes. a few loose blocks. two broken pipes. one irregular block cluster.
      entry 82: two broken pipes. two coins. two coin lines. several towers.
      entry 83: many pipes. a few upside down pipes. several loose blocks. two ascending staircases. many platforms. a few coins. one broken pipe.
      entry 87: several coins. one broken pipe. two walls. a few irregular block clusters. two descending staircases. a few enemies. a few coin lines.
      entry 89: several rectangular block clusters. two loose blocks. one broken pipe. giant gap with several chunks of floor. several irregular block clusters. a few descending staircases. one ascending staircase. many question blocks. a ton of platforms.
      entry 91: two descending staircases. a ton of loose blocks. one enemy. two broken pipes. two cannons.
      entry 93: one broken pipe. two enemies. one pipe. two descending staircases. two irregular block clusters. many coins.

  'walls'  (appears 21 times)
      entry 1: one coin line. two walls. several ascending staircases. a ton of question blocks. one coin. a few descending staircases.
      entry 2: one platform. a ton of coins. two rectangular block clusters. two irregular block clusters. many cannons. one broken pipe. giant gap with a few chunks of floor. several coin lines. two enemies. two walls.
      entry 3: a few coins. several enemies. one broken pipe. several coin lines. a few towers. one cannon. one rectangular block cluster. one descending staircase. two walls. several pipes. one upside down pipe.
      entry 5: two walls. ceiling with a few gaps. two coins. one pipe. a few loose blocks. two broken pipes. many enemies. two cannons. a few coin lines.
      entry 8: several ascending staircases. two walls. many cannons. several question blocks. ceiling with one gap. several loose blocks. a ton of pipes. a ton of coins.
      entry 13: one ascending staircase. a few pipes. two walls. a few coins. one question block. two descending staircases. a few coin lines.
      entry 15: a ton of coins. two broken pipes. many towers. a few coin lines. several rectangular block clusters. two walls.
      entry 24: two coins. several platforms. one tower. two walls. a ton of enemies.
      entry 29: a few loose blocks. two descending staircases. many pipes. many coins. giant gap with many chunks of floor. two walls. several platforms.
      entry 41: two walls. one coin.
      entry 51: two descending staircases. many coin lines. a few enemies. two walls. many question blocks. a few coins. two loose blocks. ceiling with several gaps.
      entry 52: two walls. several ascending staircases. two platforms. one irregular block cluster. many cannons. two broken pipes. a few enemies. a few coin lines. a ton of coins. two loose blocks.
      entry 54: a few rectangular block clusters. ceiling with several gaps. two walls. many pipes.
      entry 63: two walls.
      entry 65: many platforms. a ton of coins. a ton of coin lines. two walls. two broken pipes. full floor. two question blocks. one tower.
      entry 66: two ascending staircases. two walls. a ton of question blocks. ceiling with several gaps. several enemies. many cannons.
      entry 69: several ascending staircases. several descending staircases. a few question blocks. many platforms. two walls. a few loose blocks. ceiling with many gaps. many pipes.
      entry 70: several enemies. two rectangular block clusters. many coins. many platforms. two walls.
      entry 71: two walls. many question blocks. two coins. a few enemies. one broken pipe. one coin line. one loose block.
      entry 73: two walls. a ton of question blocks. two rectangular block clusters. a few descending staircases. two coins. one platform. several cannons. two coin lines.
      entry 87: several coins. one broken pipe. two walls. a few irregular block clusters. two descending staircases. a few enemies. a few coin lines.

  'wall'  (appears 18 times)
      entry 4: several enemies. one irregular block cluster. one loose block. many coin lines. one wall. many coins. a few ascending staircases.
      entry 12: one descending staircase. ceiling with several gaps. giant gap with a few chunks of floor. a few coins. one coin line. a few ascending staircases. many cannons. two broken pipes. many loose blocks. one wall.
      entry 14: one wall. one rectangular block cluster. ceiling with two gaps. a few descending staircases. two platforms. a few loose blocks. a few ascending staircases. several upside down pipes.
      entry 16: one cannon. several loose blocks. a few rectangular block clusters. one wall. two coins. many towers. many question blocks.
      entry 19: one upside down pipe. one loose block. one wall.
      entry 20: two pipes. one wall. one upside down pipe. giant gap with many chunks of floor. one broken pipe. two ascending staircases.
      entry 22: a few towers. several ascending staircases. several enemies. many upside down pipes. one wall. floor with a few gaps.
      entry 31: a few rectangular block clusters. one tower. one ascending staircase. many irregular block clusters. ceiling with many gaps. a ton of question blocks. two enemies. one wall.
      entry 35: two platforms. two irregular block clusters. floor with a ton of gaps. a few coin lines. one broken pipe. a few enemies. one cannon. one wall. two coins.
      entry 44: two irregular block clusters. a few descending staircases. full ceiling. floor with a few gaps. a few enemies. two rectangular block clusters. a few upside down pipes. two coins. one coin line. one wall.
      entry 48: several question blocks. one pipe. one wall. giant gap with a few chunks of floor.
      entry 53: a ton of pipes. a few platforms. two upside down pipes. one ascending staircase. a few question blocks. a ton of loose blocks. floor with a ton of gaps. one wall. many coins. many coin lines. two cannons.
      entry 58: several descending staircases. giant gap with many chunks of floor. two coin lines. a ton of coins. a few pipes. several platforms. one wall.
      entry 59: a ton of loose blocks. one wall. a ton of platforms.
      entry 64: one wall. many pipes. a few question blocks. several rectangular block clusters. a few descending staircases. one loose block.
      entry 79: several descending staircases. a ton of platforms. a few towers. two upside down pipes. one wall. one rectangular block cluster.
      entry 94: a ton of pipes. one wall.
      entry 95: a few question blocks. a few coins. a few irregular block clusters. two upside down pipes. several ascending staircases. a ton of enemies. giant gap with a few chunks of floor. several pipes. many loose blocks. one wall.

Verification failed: 3 illegal token(s) found across 66 occurrence(s).
```

The RandomTest dataset was the output of running `Mar1and2Mixed-data.bat`. I've also attached the file here in case the result you get differs from what is produced when running this on my home machine.

[Mar1and2_16-32-64-128_RandomTest-regular.json](https://github.com/user-attachments/files/29117840/Mar1and2_16-32-64-128_RandomTest-regular.json)

**Comment by schrum2** [comment]
I also got more unknown token errors when training with the `train-conditional-growing.bat` batch file which uses different datasets:
```
Warning: Unknown token skipped: broken
Warning: Unknown token skipped: wall
Warning: Unknown token skipped: wall
Warning: Unknown token skipped: walls
Warning: Unknown token skipped: broken
Warning: Unknown token skipped: wall
Warning: Unknown token skipped: walls
Warning: Unknown token skipped: walls
Warning: Unknown token skipped: walls
Warning: Unknown token skipped: broken
Warning: Unknown token skipped: wall
Warning: Unknown token skipped: broken
```
This is Mario again. Something clearly went wrong with the captioning at some point.

Important note: "broken" is not supposed to appear in any captions in the training set. It is also NOT supposed to appear in any captions in the RandomTest sets either. The only reason "broken" is in the vocabulary is because generated scenes can contain broken elements, and the auto-captioning identifies those and labels them as such. This is also used in computing the caption adherence score, but a caption that contains the token "broken" should never be fed to a tokenizer since "broken" should never appear in a prompt to describe an element that one desires to generate in a level.

**Comment by fwilson12** [comment]
Recent commit should finally fix the errors. Please reopen if anything else isn't working; all the bug ended up being was the presence of the 'broken' and 'wall' tokens in the caption_generator.py keyword dict. 

##
>(SURF) C:\Users\wilsonf\Documents\GitHub\MarioDiffusion>python verify_data_and_tokenizer.py --json datasets\Mar1and2_16-32-64-128_RandomTest-regular.json --pkl datasets\Mar1and2_16-32-64-128_Tokenizer-regular.pkl
Dataset:   datasets\Mar1and2_16-32-64-128_RandomTest-regular.json  (100 entries)
Tokenizer: datasets\Mar1and2_16-32-64-128_Tokenizer-regular.pkl  (vocab size 48)
Captions found: 100
Total tokens scanned: 2535
Distinct tokens:      46
SUCCESS: Every token across all captions is recognized by the tokenizer.
